apollo-compiler/src/test/graphql/com/example/antlr_tokens/TestQuery.kt

```diff
@@ -50,6 +52,9 @@ data class TestQuery(
   override fun name(): String = OPERATION_NAME
 
   override fun adapter(): ResponseAdapter<Data> = TestQuery_ResponseAdapter
+  override fun responseFields(): Map<String, Array<ResponseField>> = mapOf(
+    "" to TestQuery_ResponseAdapter.RESPONSE_FIELDS
+  )
```

Comment:
should we move this function into adapter? And we can get access to it via `Query.adapter().responseFeilds()`?

Reply:
Also this mapping is a little confusing `"" to TestQuery_ResponseAdapter.RESPONSE_FIELDS`.

I'm thinking if we can provide some abstraction for it. If I understood this right the key here is typename and the value is root field. What do you think if we introduce this abstraction:
```
sealed class ResponseFieldRoot
  class Field(val responseField: ResponseField)
  class Fragment(val possibleSubRoots: Map<String, ResponseField>)
```

Naming could be better just throwing an idea.

Reply:
>should we move this function into adapter? And we can get access to it via Query.adapter().responseFeilds()

My long term goal is actually to remove this from the adapter. Most of the contents of `ResponseField` is only used to read from the cache. For users without a cache, we can make this optional and save a bunch of bytecode and class init time.

Reply:
> the key here is typename 

Yes

> and the value is root field.

It's the list of fields that this typename will have. I like to think of this as a `FieldSet` because there's no fragment involved there.

We could introduce something like:

```
class FieldSet(val typename: String, val responseFields: Array<ResponseField>)
```

Reply:
Yeah I think that would be better.

Also maybe make it like 
```
class FieldSet(
  val responseFields: Array<ResponseField>,
  val typeCondition: String? = null
}
```
so this will indicate that this set of fields guarded by `typeCondition` if it present. 

What do you think?

Reply:
Sounds good 👍